sensors/repository_sensor.py

```diff
@@ -0,0 +1,221 @@
+import stashy
+
+from pybitbucket.auth import BasicAuthenticator
+from pybitbucket.bitbucket import Client
+from pybitbucket.ref import Branch
+from pybitbucket.user import User
+
+from timeout_decorator import timeout
+from timeout_decorator.timeout_decorator import TimeoutError
+
+from datetime import datetime
+
+from st2reactor.sensor.base import PollingSensor
+
+
+class RepositorySensor(PollingSensor):
+    DATE_FORMAT = '%Y-%m-%d %H:%M:%S'
+    TIMEOUT_SECONDS = 20
+
+    def __init__(self, sensor_service, config=None, poll_interval=None):
+        super(RepositorySensor, self).__init__(sensor_service=sensor_service,
+                                               config=config,
+                                               poll_interval=poll_interval)
+        self._trigger_ref = 'bitbucket.repository_event'
+        self._logger = self._sensor_service.get_logger(__name__)
+        self.last_commit = {}
+
+    def setup(self):
+        sensor_config = self._config.get('sensor', None)
+        if not sensor_config:
+            raise ValueError('"sensor" config value is required')
+
+        # validate the format of all repository names
+        self.repositories = sensor_config.get('repositories')
+        if not self.repositories:
+            raise ValueError('"repositories" parameter in the "sensor" is required')
+
+        if not all([len(x.split('/')) == 2 for x in self.repositories]):
+            raise ValueError('Invalid repository name is specified in the "repositories" parameter')
+
+        self.TIMEOUT_SECONDS = sensor_config.get('timeout', self.TIMEOUT_SECONDS)
+
+        # initialize global parameter
+        self.commits = {}
+
+        self.service_type = sensor_config.get('bitbucket_type')
+        if self.service_type == 'server':
+            # initialization for BitBucket Server
+            self.client = stashy.connect(sensor_config.get('bitbucket_server_url'),
+                                         self._config.get('username'),
+                                         self._config.get('password'))
+
+            self._init_server_last_commit()
+        elif self.service_type == 'cloud':
+            # initialization for BitBucket Cloud
+            self.client = Client(BasicAuthenticator(
+                self._config.get('username'),
+                self._config.get('password'),
+                self._config.get('email'),
+            ))
+            self._init_cloud_last_commit()
+        else:
+            raise ValueError('specified bitbucket type (%s) is not supported' % self.service_type)
+        self._increment_event_id()
+
+    def poll(self):
+        @timeout(self.TIMEOUT_SECONDS)
+        def get_server_updated_commits():
+            for (proj, repo) in [x.split('/') for x in self.repositories]:
```

Comment:
This would be a bit more readable if there were less levels of nesting (split it in multiple methods), but it's fine for now.